src/npyffi/mod.rs

```diff
@@ -31,19 +36,74 @@ fn get_numpy_api<'py>(
     // so we can safely cache a pointer into its interior.
     forget(capsule);
 
+    API_VERSION.get_or_init(py, || unsafe {
+        #[allow(non_snake_case)]
+        let PyArray_GetNDArrayCFeatureVersion = api.offset(211) as *const extern "C" fn() -> c_uint;
+        (*PyArray_GetNDArrayCFeatureVersion)()
+    });
+
     Ok(api)
 }
 
+const fn api_version_to_numpy_version_range(api_version: c_uint) -> (&'static str, &'static str) {
+    match api_version {
+        0..=0x00000008 => ("?", "1.7"),
+        0x00000009 => ("1.8", "1.9"),
+        0x0000000A => ("1.10", "1.12"),
+        0x0000000B => ("1.13", "1.13"),
+        0x0000000C => ("1.14", "1.15"),
+        0x0000000D => ("1.16", "1.19"),
+        0x0000000E => ("1.20", "1.21"),
+        0x0000000F => ("1.22", "1.22"),
+        0x00000010 => ("1.23", "1.24"),
+        0x00000011 => ("1.25", "1.26"),
+        0x00000012..=c_uint::MAX => ("2.0", "?"),
+    }
+}
+
 // Implements wrappers for NumPy's Array and UFunc API
 macro_rules! impl_api {
-    [$offset: expr; $fname: ident ( $($arg: ident : $t: ty),* $(,)?) $( -> $ret: ty )* ] => {
+    // API available on all versions
+    [$offset: expr; $fname: ident ($($arg: ident: $t: ty),* $(,)?) $(-> $ret: ty)?] => {
+        #[allow(non_snake_case)]
+        pub unsafe fn $fname<'py>(&self, py: Python<'py>, $($arg : $t), *) $(-> $ret)* {
+            let fptr = self.get(py, $offset) as *const extern fn ($($arg : $t), *) $(-> $ret)*;
+            (*fptr)($($arg), *)
+        }
+    };
+
+    // API with version constraints, checked at runtime
+    [$offset: expr; ..=1.26; $fname: ident ($($arg: ident: $t: ty),* $(,)?) $(-> $ret: ty)?] => {
+        impl_api![$offset; ..=0x00000011; $fname($($arg : $t), *) $(-> $ret)*];
+    };
+    [$offset: expr; 2.0..; $fname: ident ($($arg: ident: $t: ty),* $(,)?) $(-> $ret: ty)?] => {
+        impl_api![$offset; 0x00000012..; $fname($($arg : $t), *) $(-> $ret)*];
+    };
```

Comment:
The pattern syntax is nice, but are any but the two NumPy 1 and NumPy 2 variants actually used?

If not, please simplify this to have exactly two patterns here, one for NumPy 1 implementing the maximum version of 0x11` and one for NumPy implementing the minimum version of 0x12`.

Reply:
(I think this would also imply that we can loose `api_version_to_numpy_version_range` as we really only need to say "requires NumPy 1/2" which is a nice follow-up simplification.)

Reply:
Done in #442.